src/platform/linux/portalgrab.cpp

```diff
@@ -0,0 +1,777 @@
+      struct spa_pod_builder pod_builder = SPA_POD_BUILDER_INIT(buffer, sizeof(buffer));
+      params[n_params++] = (const struct spa_pod *) spa_pod_builder_add_object(&pod_builder, SPA_TYPE_OBJECT_ParamBuffers, SPA_PARAM_Buffers, SPA_PARAM_BUFFERS_dataType, SPA_POD_Int(buffer_types));
+      pw_stream_update_params(d->stream, params, n_params);
+    }
+
+    constexpr static const struct pw_stream_events stream_events = {
+      .version = PW_VERSION_STREAM_EVENTS,
+      .param_changed = on_param_changed,
+      .process = on_process,
+    };
+  };
+
+  class portal_t: public platf::display_t {
+  public:
+    int
+    init(platf::mem_type_e hwdevice_type, const std::string &display_name, const ::video::config_t &config) {
+      framerate = config.framerate;
+      delay = std::chrono::nanoseconds { 1s } / framerate;
+      mem_type = hwdevice_type;
+
+      if (get_dmabuf_modifiers() < 0)
+        return -1;
+      if (dbus.init() < 0)
+        return -1;
+      if (dbus.connect_to_portal() < 0)
+        return -1;
+
+      width = dbus.width;
+      height = dbus.height;
+      framerate = config.framerate;
+
+      pipewire.init(dbus.pipewire_fd, dbus.pipewire_node);
+
+      return 0;
+    }
+
+    platf::capture_e
+    snapshot(const pull_free_image_cb_t &pull_free_image_cb, std::shared_ptr<platf::img_t> &img_out, std::chrono::milliseconds timeout, bool show_cursor) {
+      // FIXME: show_cursor is ignored
+      if (!pull_free_image_cb(img_out)) {
+        return platf::capture_e::interrupted;
+      }
+
+      auto img_egl = (egl::img_descriptor_t *) img_out.get();
+      img_egl->reset();
+      pipewire.fill_img(img_egl);
+      img_egl->sequence = ++sequence;
+
+      return platf::capture_e::ok;
+    }
+
+    std::shared_ptr<platf::img_t>
+    alloc_img() override {
+      // Note: this img_t type is also used for memory buffers
+      auto img = std::make_shared<egl::img_descriptor_t>();
+
+      img->width = width;
+      img->height = height;
+      img->pixel_pitch = 4;
+      img->row_pitch = img->pixel_pitch * width;
+      img->sequence = 0;
+      img->serial = std::numeric_limits<decltype(img->serial)>::max();
+      img->data = nullptr;
+      std::fill_n(img->sd.fds, 4, -1);
+
+      return img;
+    }
+
+    platf::capture_e
+    capture(const push_captured_image_cb_t &push_captured_image_cb, const pull_free_image_cb_t &pull_free_image_cb, bool *cursor) override {
+      auto next_frame = std::chrono::steady_clock::now();
+
+      pipewire.ensure_stream(mem_type, width, height, framerate, (struct dmabuf_format_info_t*) dmabuf_infos, n_dmabuf_infos);
+
+      while (true) {
+        auto now = std::chrono::steady_clock::now();
+
+        if (next_frame > now) {
+          std::this_thread::sleep_for((next_frame - now) / 3 * 2);
+        }
+        while (next_frame > now) {
+          std::this_thread::sleep_for(1ns);
+          now = std::chrono::steady_clock::now();
+        }
+        next_frame = now + delay;
```

Comment:
For consistency you might want to incorporate the changes from #2333.
The changes would be fairly benign: kick the two-step sleep; base the next_frame time on previous theoretical instead of after capture; add sleep overshoot tracking.

Reply:
Cheers! Will look to add this.